core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/AnonymizedNodeRepresentation.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.trino.sql.planner.planprinter.anonymize;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.JsonSubTypes;
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
+import io.trino.sql.planner.plan.PlanNodeId;
+
+import java.util.List;
+
+import static java.util.Objects.requireNonNull;
+
+@JsonTypeInfo(
+        use = JsonTypeInfo.Id.NAME,
+        property = "@type")
+@JsonSubTypes({
+        @JsonSubTypes.Type(value = DefaultNodeRepresentation.class, name = "default"),
+        @JsonSubTypes.Type(value = ProjectNodeRepresentation.class, name = "project"),
+        @JsonSubTypes.Type(value = TableScanNodeRepresentation.class, name = "tableScan"),
+        @JsonSubTypes.Type(value = ValuesNodeRepresentation.class, name = "values"),
+        @JsonSubTypes.Type(value = AggregationNodeRepresentation.class, name = "aggregation"),
+        @JsonSubTypes.Type(value = FilterNodeRepresentation.class, name = "filter"),
+        @JsonSubTypes.Type(value = WindowNodeRepresentation.class, name = "window"),
+        @JsonSubTypes.Type(value = TopNNodeRepresentation.class, name = "topN"),
+        @JsonSubTypes.Type(value = SortNodeRepresentation.class, name = "sort"),
+        @JsonSubTypes.Type(value = RemoteSourceNodeRepresentation.class, name = "remoteSource"),
+        @JsonSubTypes.Type(value = ExchangeNodeRepresentation.class, name = "exchange"),
+        @JsonSubTypes.Type(value = JoinNodeRepresentation.class, name = "join"),
+        @JsonSubTypes.Type(value = SemiJoinNodeRepresentation.class, name = "semiJoin"),
+})
+public abstract class AnonymizedNodeRepresentation
```

Comment:
Why do we need a separate hierarchy for the anonymized representation? It should be able to encode the same information in a regular plan representation, with all sensitive information masked or removed

Reply:
There are few problems wrt simply anonymizing each `PlanNode` and expose the tree as JSON.

1. In case, any new field is added to any `PlanNode`, which contains some sensitive information, we could leak that information. So, we would always have to remember to update this code whenever any plan node changes.
2. Another issue is that it seems like we can't serialize `PlanNode` into JSON because it contains different expressions which don't implement Jackson annotations.

And if we go with the `jsonFragmentPlan` route in the `PlanPrinter`, which the above PR kind of followed. Then, the issue is that the result JSON contains all the specific plan node fields under the `details` key in a text format that is hard to parse and subject to change.

For instance:
```
{
    "id" : "876",
    "name" : "InnerJoin",
    "identifier" : "[(\"orderkey_7\" = \"orderkey_4\")][$hashvalue_38, $hashvalue_39]",
    "details" : "Distribution: REPLICATED\ndynamicFilterAssignments = {orderkey_4 -> #df_1052}",
    ...
}
```

cc @raunaqmorarka @sopel39

Reply:
I'm not talking about anonymizing PlanNodes. That's the wrong approach. However, we could have a converter from PlanNode -> PlanRepresentation that is designed specifically to produce anonymized information. It would not suffer from the issue you describe because it would be a "whitelist" of conversions.

Reply:
`NodeRepresentation` in `PlanRepresentation` forces all the node level details into a single `details` string that isn't guaranteed to be in any particular format. For our use case we want the details to be easily parseable by some downstream consumer of event listener.
I suppose we could stuff all the anonymized details into a map and convert that into a json string. But that seems an unnatural approach compared to serialization of representation objects we've done here.
It's also odd behaviour for `EXPLAIN (TYPE LOGICAL, FORMAT JSON)` to give up on json format and start printing unformatted lines for the `details` part of query plans.
How about we extend `NodeRepresentation` to be a class hierarchy like we have here ? That would allow `EXPLAIN (TYPE LOGICAL, FORMAT JSON)` to print json all the way and the plan anonymiser we have here would be able to reuse the same hierarchy to populate anonymized info.